wordpress-mobile/WordPress-iOS
# [Tooling] Migrate download_metadata.swift to gp_downloadmetadata action

This work is part of paaHJt-2Ib-p2 — task [i2]

⚠️  This PR targets the `release/19.7` branch, as this is where it's the most likely that the changes from this one and the relevant lanes will be tested next.

### Why / How

As an effort to migrate all our legacy Localization tooling and scripts into our release-toolkit fastlane plugin, this PR:
 - Removes the `fastlane/download_metadata.swift` script and replace it with a call to our existing `gp_downloadmetadata` action
 - Moves the metadata files that were in `fastlane/appstoreres/{metadata,jetpack_metadata}/source/{description,keywords,name,subtitle}.txt` into `fastlane/{metadata,jetpack_metadata}/default` instead, for consistency.
 - Removes the `fastlane/{metadata,jetpack_metadata}/en-US/{description,keywords,name,subtitle,release_notes}.txt` files (and add a `.gitkeep` file in `en-US` dir to still keep it around) so that `default/` is the new source of truth (used both for `en-US` and for any other locales that would not have translations available in GlotPress for those keys)
   - The `default/` pseudo-locale folder is now the source of truth for the originals of the metadata copies that we translate and include in our translation pipeline and automation, and in theory should act as a fallback for locales missing translations(†).
   - Some `.txt` files are still present in `en-US` for now though. Those are the files that haven't been configured to be included in our translation pipeline yet (files that were not treated by `download_metadata.swift` and not listed in `APPSTORECONNECT_METADATA_KEYS_MAP` yet). Those copies (`apple_tv_privacy_policy`, `marketting_url`, …) change rarely (if ever?), and in the current state of the repo all seem duplicated in the other locale folders (and our script never touch those — they seem to always been manually updated apparently). In a future PR, we might consider either adding those extra files to the pipeline (i.e. to the `AppStoreStrings.po` file / `ios_update_metadata_source` call, and in the `APPSTORECONNECT_METADATA_KEYS_MAP` map)… or keeping them untranslated and out of our translation pipeline/automation — but in that case we ought to delete all duplicate `.txt` files with English copy from each locale folder, and move those `.txt` files to `default/` for a single source of truth. All this seems outside of the scope of this PR though, so I preferred to keep that for later to keep this PR focused on porting the existing configuration only.
 - Make the lanes copy the `WordPress/{Jetpack,}/Resources/release_notes.txt` original copy into 
`fastlane/{metadata,jetpack_metadata}/default/release_notes.txt` at the same time that it download the translations for other locales from GlotPress, so that the Editorial Copy for the release notes is available in the path expected by `deliver`
    - In the future we hope to make that `fastlane/{metadata,jetpack_metadata}/default/release_notes.txt` path the official source of truth that we'll use as part of our Release Scenarios and the automation for the Announcement P2 post etc. But this is work for a future iteration that felt outside of the scope of this project.

## To Test

 - 🕵️  Check the code logic and the map of locales matches what was in the deleted `download_metadata.swift` — to ensure we still translate the same locales
 - ✂️  Cut a test branch from this one
 - 😈  Remove some metadata folders like `fastlane/jetpack_metadata/fr` and/or `fastlane/metadata/da` (so that we can check that they'll be re-created, even if empty) and commit.
 - 💻  Run `bundle exec fastlane download_wordpress_localized_app_store_metadata` then `bundle exec fastlane download_jetpack_localized_app_store_metadata` 
 - 🕵️  Check that the action worked and created 2 commits, one for WordPress and one for Jetpack
 - 🕵️  Check the content of the commits:
   - The `{description,keywords,name,subtitle}.txt` files in both `fastlane/metadata/*/` and `fastlane/metadata/*/` were unchanged (modulo CRLF at end of file which can be ignored) — since it's unlikely that new translations for those came in since last time we updated them.
   - The folders for locales you deleted in the 😈  step got recreated and re-populated with the proper copies
   - The copies for locales which didn't have any translations in GlotPress (e.g. `da`, `en-AU`, …) got deleted(‡), but their corresponding folder still got created with a `.gitkeep` file, even if no translation / `*.txt` files were found for that locale
   - The `release_notes.txt` files got copied from `WordPress/{Jetpack,}/Resources/release_notes.txt` into `fastlane/{metadata,jetpack_metadata}/default/` for WP and JP
 - 🧹 Delete your test branch

## Footnotes

_(†) It will probably be worth checking during the next AppStore submission that this indeed works as expected with `deliver` — especially check locales missing some translations for (some of all of) their metadata still got updated, but with the fallback copy from `default/`. Especially thinking about cases like `metadata/da/name.txt` which was still `"WordPress"` from a previous original copy while the new copy was never translated; with the changes from this PR, `deliever` should update the `da` locale in ASC and fallback to `metadata/default/name.txt` (aka `"WordPress – Website Builder"`) for the copy used._

_(‡) This shows that this PR fixed a bug where the previous `download_metadata.swift` failed to delete copies missing a translation… which led to old copies being used if a translation for new ones were not available in GlotPress… while the new setup ensures the folder exists but is empty for those cases, which should make `deliver` still update those locales but fallback to the copies in `default/` for those._

## What's next

Once a similar change has been adopted in all client apps (i.e. being ported WCiOS if relevant), we should consider removing the legacy methods relying on that script in the release-toolkit:

 - remove the `skip_glotpress` parameter (and associated code using it) in `ios_bump_version_release` logic to stop doing the updates at all (to do the same behavior as if the parameter was always `false`)
 - remove the `include_metadata:` parameter from `Ios::GitHelper::commit_version_bump` as well

We could at least mark them as explicitly deprecated in the next minor release, and remove them in the next major release.
